src/lua/LuaSerializer.cpp

```diff
@@ -359,12 +367,85 @@ void LuaSerializer::UninitTableRefs()
 	lua_setfield(l, LUA_REGISTRYINDEX, "PiSerializer");
 
 	lua_pushnil(l);
-	lua_setfield(l, LUA_REGISTRYINDEX, "PiSerializerTableRefs");
+	lua_setfield(l, LUA_REGISTRYINDEX, NS_REFTABLE);
 
 	lua_pushnil(l);
 	lua_setfield(l, LUA_REGISTRYINDEX, "PiLuaRefLoadTable");
 }
 
+void LuaSerializer::SavePersistent(Json &json)
+{
+	lua_State *l = Lua::manager->GetLuaState();
+	LUA_DEBUG_START(l);
+
+	// NOTE: this must be an array for consistent deserialization order
+	Json persist = Json::array();
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_PERSISTENT);
+	lua_pushnil(l);
+
+	while (lua_next(l, -2)) {
+		Json entry = Json::object();
+
+		// persistent, id, value
+		std::string id = LuaPull<std::string>(l, -2);
+
+		// Serialize all registered persistent objects as a "backup" version
+		// (This will populate the PiSerializerTableRefs cache)
+		entry["persistId"] = id;
+		pickle_json(l, -1, entry, id);
+
+		persist.push_back(std::move(entry));
+
+		lua_pop(l, 1);
+	}
+
+	lua_pop(l, 1);
+
+	json["lua_persistent_json"] = std::move(persist);
+
+	LUA_DEBUG_END(l, 0);
+}
+
+void LuaSerializer::LoadPersistent(const Json &json)
+{
+	lua_State *l = Lua::manager->GetLuaState();
+	LUA_DEBUG_START(l);
+
+	const Json &persist = json["lua_persistent_json"];
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_REFTABLE);
+	int idx_reftable = lua_gettop(l);
+
+	luaL_getsubtable(l, LUA_REGISTRYINDEX, NS_PERSISTENT);
+
+	for (const auto &item : persist) {
+		std::string id = item["persistId"].get<std::string>();
+
+		lua_pushinteger(l, item["ref"].get<lua_Integer>());
+		lua_getfield(l, -2, id.c_str()); // reftable, persist, ptr, pvalue
+
+		// We must unpickle the saved persistent object, because it may contain
+		// "trapped" serialized representations of non-persistent tables
+		unpickle_json(l, item);
+
+		// No valid persistent value (mismatch in serialization!)
+		if (lua_isnil(l, -2)) {
+			Log::Warning("Restoring missing persistent object '{}' from savefile", id);
+			lua_replace(l, -2); // reftable, persistent, ptr, svalue
+		} else {
+			lua_pop(l, 1); // reftable, persistent, ptr, pvalue
+		}
+
+		// All references to the prior saved value are replaced with the persistent object
+		lua_settable(l, idx_reftable);
+	}
```

Comment:
It seems that the missing persistent object is not put into the NS_PERSISTENT table, and therefore will not be saved to the next save? So the next time we load we will still get a dangling object referencing it?

Reply:
*Good* catch, thank you! I had completely overlooked this. You are correct, this should probably be written back to the persistent object cache. However, modifying the persistent object cache in this way will affect other saves made after starting a new game during the same session, so it's not the cleanest solution.

The latter issue can be resolved by reloading the Lua environment between game starts as we are (eventually) planning on doing, so it's not a huge issue at the moment.